### DISCUSSION TITLE
Maven repository url's incorrectly marked as infinitely recursive

### DISCUSSION THREAD

**Discussion by Orbifoldt** [renovatebot/renovate#24815] [Suggest an Idea]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

When defining custom repositories in the maven ```pom.xml``` whose urls use a nested property as shown below, Renovate will skip all dependencies, giving the ```recursive-placeholder```-reason as root cause. See the debug logs below.

Expected behaviour: it should be able to just resolve the urls as they are not infinitely recursive. 

Occurence: I first noticed this issue in a self-hosted instance of Renovate (using latest at time of writing). I was able to reproduce the issue with the Mend Renovate host too. Here an excerpt from the  ```pom.xml``` from https://github.com/Orbifoldt/renovate-recursive-repos-bug :
```xml
  <properties>
    <url.repo.base>https://my.repo.com</url.repo.base>
    <url.repo.releases>${url.repo.base}/content/repositories/java-releases/</url.repo.releases>
  </properties>

  <repositories>
    <repository>
      <id>repo-dependencies</id>
      <url>${url.repo.base}/content/groups/java-group/</url>
    </repository>
    <repository>
      <id>repo-releases</id>
      <url>${url.repo.releases}</url>
    </repository>
  </repositories>
```
Note that here the property ```${url.repo.releases}``` uses the property ```${url.repo.base}```, I assume that this dependency (which is not an infinite recursion) is causing the issue. 

### Relevant debug logs

<details><summary>Logs</summary>
Log from renovate run #690915187 

```
DEBUG: packageFiles with updates
{
  "config": {
    "maven": [
      {
        "datasource": "maven",
        "packageFile": "pom.xml",
        "deps": [
          {
            "datasource": "maven",
            "depName": "org.apache.tika:tika-core",
            "currentValue": "1.20",
            "fileReplacePosition": 848,
            "registryUrls": [
              "https://repo.maven.apache.org/maven2",
              "${url.repo.base}/content/groups/java-group/",
              "${url.repo.releases}"
            ],
            "depType": "compile",
            "skipReason": "recursive-placeholder",
            "depIndex": 0,
            "updates": []
          }
        ],
        "packageFileVersion": "1.0-SNAPSHOT"
      }
    ]
  }
}
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

**Comment by viceice** [comment]
it's not a bug. we don't resolve those placeholders

**Comment by Orbifoldt** [comment]
Is there a specific reason it is not supported? 
Perhaps the skip reason could at least be made a bit more insightful? It took me a while to figure out that the repositories were causing the issue

**Comment by rarkins** [comment]
I'm not sure I fully understand the problem, although naturally it's good to have all errors easy to understand. Is there a benefit to performing this resolution instead of skipping it like today? e.g. is there a PR which could be raised against the repo if so?

**Comment by Orbifoldt** [comment]
The main issue for me is it causes renovate to completely skip all dependency checks for the whole project. This is because if you define a repository in the pom it's applied to everything. So none of the dependencies are checked, even if you overwrite the dependcy urls in the renovate config, and hence no PRs are raised ever.

I am not sure what would be a nicer way to deal with this. Perhaps it's possible to just skip all unresolvable urls. Then it would only skip the dependency if there are no resolved urls left? In the log I shared it displayed:
```
            "registryUrls": [
              "https://repo.maven.apache.org/maven2",
              "${url.repo.base}/content/groups/java-group/",
              "${url.repo.releases}"
            ],
```
So here it could at least try maven central. Then using a host rule you can overwrite it and point to the correct (private) repo? What do you think @rarkins ?

**Comment by rarkins** [comment]
The datasource layer should skip any invalid registry URL, such as the two shown above. So maybe we can remove the `skipReason` in this case and let the datasource handle it (or let the user override it with `packageRules`).
